File: csharp/ShrewdEvolver.UnitTests/LogMessageTemplateTest.cs
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Xunit;
using static AaronicSubstances.ShrewdEvolver.LogMessageTemplateParser;
using static AaronicSubstances.ShrewdEvolver.UnitTests.TestUtils;

namespace AaronicSubstances.ShrewdEvolver.UnitTests
{
    public class LogMessageTemplateTest
    {
        public class EmbeddableLogRecordImpl: LogMessageTemplate
        {

            public EmbeddableLogRecordImpl(string formatString, object treeData, List<object> positionalArgs):
                base(formatString, treeData, positionalArgs)
            { }

            protected override string SerializeData(object treeDataSlice)
            {
                return JsonConvert.SerializeObject(treeDataSlice);
            }
        }

        public class TestLogRecordImp2 : LogMessageTemplate
        {

            public TestLogRecordImp2(string formatString, object treeData, List<object> positionalArgs) :
                base(formatString, treeData, positionalArgs)
            { }

            protected override object HandleNonExistentTreeDataSlice(PartDescriptor part, int nonExistentIndex)
            {
                return null;
            }

            protected override object HandleNonExistentPositionalArg(PartDescriptor part)
            {
                return null;
            }
        }

        [Theory]
        [MemberData(nameof(CreatTestGetTreeDataSliceData))]
        public void TestGetTreeDataSlice(object treeData, List<object> treeDataKey, object expected)
        {
            LogMessageTemplate instance = new TestLogRecordImp2("", treeData, null);
            var part = new PartDescriptor(0, 0, treeDataKey);
            object actual = instance.GetTreeDataSlice(treeData, part);
            Assert.Equal(actual, expected);
        }

        public static List<object[]> CreatTestGetTreeDataSliceData()
        {
            return new List<object[]>
            {
                new object[]{ null, ToList(), null },
                new object[]{ "", ToList(), "" },
                new object[]{ 2, ToList(), 2 },
                new object[]{ ToList(2), ToList("f"), null },
                new object[]{ ToList(2), ToList(0), 2 },
                new object[]{ ToList(21, ToList()), ToList(10), null },
                new object[]{ ToList(21, ToList()), ToList(0), 21 },
                new object[]{ ToList(21, ToList()), ToList(1), ToList() },
                new object[]{ ToList(21, ToList()), ToList(-1), ToList() },
                new object[]{ ToList(21, ToList()), ToList(-2), null },
                new object[]{ ToDict("a", 1), ToList(), ToDict("a", 1) },
                new object[]{ ToDict("a", 1), ToList(0), null },
                new object[]{ ToDict("a", 1), ToList("a"), 1 },
                new object[]{ ToDict("a", 1, "b", 2), ToList("b"), 2 },
                new object[]{ ToDict("a", 1, "b", 2), ToList("c"), null },
                new object[]{ ToDict("a", 1, "b", ToList("e", true)), ToList("b", 0), "e" },
                new object[]{ ToDict("a", 1, "b", ToList("e", true)), ToList("b", 1), true },
                new object[]{ ToDict("a", 1, "b", ToDict("e", true)), ToList("b"), ToDict("e", true) },
                new object[]{ ToDict("a", 1, "b", ToDict("e", true)), ToList("b", "e"), true }
            };
        }

        [Theory]
        [MemberData(nameof(CreateTestGetPositionalArgData))]
        public void TestGetPositionalArg(List<object> args, int index, object expected)
        {
            LogMessageTemplate instance = new TestLogRecordImp2("", null, args);
            var part = new PartDescriptor(0, 0, index);
            object actual = instance.GetPositionalArg(args, part);
            Assert.Equal(actual, expected);
        }

        public static List<object[]> CreateTestGetPositionalArgData()
        {
            return new List<object[]>
            {
                new object[]{ null, 0, null },
                new object[]{ ToList(), 0, null },
                new object[]{ ToList(), -1, null },
                new object[]{ ToList(), 1, null },
                new object[]{ ToList(1), 0,  1 },
                new object[]{ ToList(1), 1,  null },
                new object[]{ ToList(1), -1, null },
                new object[]{ ToList(1), -2, null },
                new object[]{ ToList(1, 2), 0,  1 },
                new object[]{ ToList(1, 2), 1,  2 },
                new object[]{ ToList(1, 2), -1,  2 },
                new object[]{ ToList(1, 2), 2,  null },
                new object[]{ ToList(1, 2), -2,  null },
                new object[]{ ToList(1, 2), -20,  null }
            };
        }

        [Theory]
        [MemberData(nameof(CreateTestToUnstructuredLogRecordData))]
        public void TestToUnstructuredLogRecord(string messageTemplate, object keywordArgs,
            List<object> positionalArgs,
            string expectedFormat, List<string> expectedFormatArgs)
        {
            LogMessageTemplate instance = new EmbeddableLogRecordImpl(messageTemplate, keywordArgs,
                positionalArgs);
            LogMessageTemplate.Unstructured output = instance.ToUnstructuredLogRecord();
            string actualFormat = output.FormatString;
            Assert.Equal(actualFormat, expectedFormat);
            List<string> actualFormatArgs = output.FormatArgs.Select(x => x.ToString()).ToList();
            Assert.Equal(new CollectionWrapper(actualFormatArgs), new CollectionWrapper(expectedFormatArgs));
        }

        public static List<object[]> CreateTestToUnstructuredLogRecordData()
        {
            return new List<object[]>
            {
                new object[]{ "", ToDict(), ToList(), "", ToGenericList<string>() },
                new object[]{ "{a}{0}", ToDict("a", "yes"), ToList(1), "{0}{1}", ToGenericList("\"yes\"", "1") },
                new object[]{ "{}{0}", ToDict("a", "yes"), ToList("1"), "{0}{1}", ToGenericList("{\"a\":\"yes\"}", "1") },
                new object[]{ "{$}{$0}", new CollectionWrapper(ToDict("a", "yes")), ToList("1"), "{0}{1}", ToGenericList("{a=yes}", "1") },
                new object[]{ "{@}{@0}", ToDict("a", "yes"), ToList("1"), "{0}{1}", ToGenericList("{\"a\":\"yes\"}", "\"1\"") }
            };
        }

        [Theory]
        [MemberData(nameof(CreateTestToStringData))]
        public void TestToString(string messageTemplate, object keywordArgs,
            List<object> positionalArgs, string expected)
        {
            LogMessageTemplate instance = new EmbeddableLogRecordImpl(messageTemplate, keywordArgs,
                positionalArgs);
            string actual = instance.ToString();
            Assert.Equal(actual, expected);
        }

        public static List<object[]> CreateTestToStringData()
        {
            return new List<object[]>
            {
                new object[]{ "", ToDict(), ToList(), "" },
                new object[]{ "{a}{0}", ToDict("a", "yes"), ToList(1), "\"yes\"1" },
                new object[]{ "{}{0}", ToDict("a", "yes"), ToList("1"), "{\"a\":\"yes\"}1" },
                new object[]{ "{$}{$0}", new CollectionWrapper(ToDict("a", "yes")), ToList("1"), "{a=yes}1" },
                new object[]{ "{@}{@0}", ToDict("a", "yes"), ToList("1"), "{\"a\":\"yes\"}\"1\"" },
                new object[]{ "{@}{@0}{8}", ToDict("a", "yes"), ToList("1"), "{\"a\":\"yes\"}\"1\"{8}" }
            };
        }
    }
}

File: java/src/test/java/com/aaronicsubstances/shrewd/evolver/LogMessageTemplateTest.java
package com.aaronicsubstances.shrewd.evolver;

import java.util.List;
import java.util.stream.Collectors;

import com.aaronicsubstances.shrewd.evolver.LogMessageTemplateParser.PartDescriptor;

import com.google.gson.Gson;

import static java.util.Arrays.asList;
import static com.aaronicsubstances.shrewd.evolver.TestUtils.toMap;

import org.testng.annotations.DataProvider;
import org.testng.annotations.Test;

import static org.testng.Assert.*;

public class LogMessageTemplateTest {

    public static class EmbeddableLogRecordImpl extends LogMessageTemplate {

        public EmbeddableLogRecordImpl(String formatString, Object treeData, List<Object> positionalArgs) {
            super(formatString, treeData, positionalArgs);
        }
		
		@Override
		protected String serializeData(Object treeDataSlice) {
			return new Gson().toJson(treeDataSlice);
		}

        @Override
        protected String escapeLiteral(String literal, boolean forLogger) {
            if (forLogger) {
				// Assume SLF4J
                return literal.replace("{", "\\{");
            }
            return super.escapeLiteral(literal, forLogger);
        }

        @Override
        protected String generatePositionIndicator(int position, boolean forLogger) {
            if (forLogger) {
				// Assume SLF4J
                return "{}";
            }
            return super.generatePositionIndicator(position, forLogger);
        }
    }
    
    @Test(dataProvider = "creatTestGetTreeDataSliceData")
    public void testGetTreeDataSlice(Object treeData, List<Object> treeDataKey, Object expected) {
        LogMessageTemplate instance = new EmbeddableLogRecordImpl("", treeData, null) {
            
            @Override
            protected Object handleNonExistentTreeDataSlice(PartDescriptor part, int nonExistentIndex) {
                return null;
            }
        };
        PartDescriptor part = new PartDescriptor(0, 0, treeDataKey);
        Object actual = instance.getTreeDataSlice(treeData, part);
        assertEquals(actual, expected);
    }

    @DataProvider
    public Object[][] creatTestGetTreeDataSliceData() {
        return new Object[][]{
            { null, asList(), null },
            { "", asList(), "" },
            { 2, asList(), 2 },
            { asList(2), asList("f"), null },
            { asList(2), asList(0), 2 },
            { asList(21, asList()), asList(10), null },
            { asList(21, asList()), asList(0), 21 },
            { asList(21, asList()), asList(1), asList() },
            { asList(21, asList()), asList(-1), asList() },
            { asList(21, asList()), asList(-2), null },
            { toMap("a", 1), asList(), toMap("a", 1) },
            { toMap("a", 1), asList(0), null },
            { toMap("a", 1), asList("a"), 1 },
            { toMap("a", 1, "b", 2), asList("b"), 2 },
            { toMap("a", 1, "b", 2), asList("c"), null },
            { toMap("a", 1, "b", asList("e", true)), asList("b", 0), "e" },
            { toMap("a", 1, "b", asList("e", true)), asList("b", 1), true },
            { toMap("a", 1, "b", toMap("e", true)), asList("b"), toMap("e", true) },
            { toMap("a", 1, "b", toMap("e", true)), asList("b", "e"), true }
        };
    }

    @Test(dataProvider = "createTestGetPositionalArgData")
    public void testGetPositionalArg(List<Object> args, int index, Object expected) {
        LogMessageTemplate instance = new EmbeddableLogRecordImpl("", null, args) {
            
            @Override
            protected Object handleNonExistentPositionalArg(PartDescriptor part) {
                return null;
            }
        };
        PartDescriptor part = new PartDescriptor(0, 0, index);
        Object actual = instance.getPositionalArg(args, part);
        assertEquals(actual, expected);
    }

    @DataProvider
    public Object[][] createTestGetPositionalArgData() {
        return new Object[][]{
            { null, 0, null },
            { asList(), 0, null },
            { asList(), -1, null },
            { asList(), 1, null },
            { asList(1), 0,  1 },
            { asList(1), 1,  null },
            { asList(1), -1, null },
            { asList(1), -2, null },
            { asList(1, 2), 0,  1 },
            { asList(1, 2), 1,  2 },
            { asList(1, 2), -1,  2 },
            { asList(1, 2), 2,  null },
            { asList(1, 2), -2,  null },
            { asList(1, 2), -20,  null }
        };
    }

    @Test(dataProvider = "createTestToUnstructuredLogRecordData")
    public void testToUnstructuredLogRecord(String messageTemplate, Object keywordArgs,
            List<Object> positionalArgs,
            String expectedFormat, List<String> expectedFormatArgs) {
        LogMessageTemplate instance = new EmbeddableLogRecordImpl(messageTemplate, keywordArgs,
            positionalArgs);
        LogMessageTemplate.Unstructured output = instance.toUnstructuredLogRecord();
        String actualFormat = output.getFormatString();
        assertEquals(actualFormat, expectedFormat);
        List<String> actualFormatArgs = output.getFormatArgs().stream()
            .map(x -> x.toString()).collect(Collectors.toList());
        assertEquals(actualFormatArgs, expectedFormatArgs);
    }

    @DataProvider
    public Object[][] createTestToUnstructuredLogRecordData() {
        return new Object[][]{
            { "", toMap(), asList(), "", asList() },
            { "{a}{0}", toMap("a", "yes"), asList(1), "{}{}", asList("\"yes\"", "1") },
            { "{}{0}", toMap("a", "yes"), asList("1"), "{}{}", asList("{\"a\":\"yes\"}", "1") },
            { "{$}{$0}", toMap("a", "yes"), asList("1"), "{}{}", asList("{a=yes}", "1") },
            { "{@}{@0}", toMap("a", "yes"), asList("1"), "{}{}", asList("{\"a\":\"yes\"}", "\"1\"") }
        };
    }

    @Test(dataProvider = "createTestToStructuredLogRecordData")
    public void testToStructuredLogRecord(String messageTemplate, Object keywordArgs,
            List<Object> positionalArgs, String expected) {
        LogMessageTemplate instance = new EmbeddableLogRecordImpl(messageTemplate, keywordArgs,
            positionalArgs);
        Object actual = instance.toStructuredLogRecord().toString();
        assertEquals(actual, expected);
    }

    @DataProvider
    public Object[][] createTestToStructuredLogRecordData() {
        return new Object[][]{
            { "", toMap(), asList(), "{}" },
            { "{a}{0}", toMap("a", "yes"), null, "{\"a\":\"yes\"}" },
            { "{}{0}", toMap("a", "yes", "b", 2), null, "{\"a\":\"yes\",\"b\":2}" },
            { "{$}{$0}", asList("a", "yes"), null, "[\"a\",\"yes\"]" },
            { "{@}{@0}", "yes", null, "\"yes\"" }
        };
    }

    @Test(dataProvider = "createTestToStringData")
    public void testToString(String messageTemplate, Object keywordArgs,
            List<Object> positionalArgs, String expected) {
        LogMessageTemplate instance = new EmbeddableLogRecordImpl(messageTemplate, keywordArgs,
            positionalArgs);
        String actual = instance.toString();
        assertEquals(actual, expected);
    }

    @DataProvider
    public Object[][] createTestToStringData() {
        return new Object[][]{
            { "", toMap(), asList(), "" },
            { "{a}{0}", toMap("a", "yes"), asList(1), "\"yes\"1" },
            { "{}{0}", toMap("a", "yes"), asList("1"), "{\"a\":\"yes\"}1" },
            { "{$}{$0}", toMap("a", "yes"), asList("1"), "{a=yes}1" },
            { "{@}{@0}", toMap("a", "yes"), asList("1"), "{\"a\":\"yes\"}\"1\"" },
            { "{@}{@0}{8}", toMap("a", "yes"), asList("1"), "{\"a\":\"yes\"}\"1\"{8}" }
        };
    }
}